Add config handler

abhijeetkaur opened the pull request:
Add "service_config" field in Service class that stores the service
related configurations.

Tested the working using yoda bot (since yoda bot's testing is
complete, added it to embedded bots registry).

Modifications also made to use utility functions from zulip_bot/lib.py
like 'initialize_config' for bots using additional configurations (which
till now was not getting called) and to check if bot was mentioned and
remove the at-mention bot name from the message content.

Only the last commit is relevant for this PR, the first 3 commits are from the PR opened on July 16th #5811. #5811 should be merged before this. 
Also, this PR depends on https://github.com/zulip/python-zulip-api/pull/64 to be merged and released.

In my local env, I used the following commands to run this 
`
pip uninstall zulip_bots`
 `pip install -e "git+https://github.com/abhijeetkaur/python-zulip-api.git@run-embedded-bots#egg=zulip_bots==0.3.2&subdirectory=zulip_bots"
`

abhijeetkaur commented:
TODOs:
* We need to add a test file to validate the input that the user enters as bot-configuration.
* UI to add configuration for embedded bots.

timabbott commented on zerver/lib/bot_lib.py:
```diff
@@ -70,11 +69,9 @@ def send_reply(self, message, response):
+        # type: (str) -> Dict[str, Any]
+        service_profile = get_service_profile(user_profile_id=self.user_profile.id, service_name=bot_name)
+
+        config_dict = ast.literal_eval(service_profile.service_config)
```
We definitely don't want to do this; it's a significant security problem to run user-provided code in the Zulip Python context.  Why not just have the config be stored in JSON format and use `ujson.loads()`?  That's safe.

timabbott commented on zerver/models.py:
```diff
@@ -1846,6 +1846,9 @@ class Service(models.Model):
     }  # type: Dict[int, Text]
 
+    # Contains YAML-format text
+    service_config = models.TextField(default=u'{}')  # type: Text
```
This needs a database migration.

abhijeetkaur commented:
This PR will fail until zulip works off zulip_bots from https://github.com/zulip/python-zulip-api/pull/64

derAnfaenger commented on static/js/settings_bots.js:
```diff
@@ -27,6 +27,8 @@ exports.type_id_to_string = function (type_id) {
     } else if (type_id === 3) {
         return i18n.t("Outgoing webhook");
+    } else if (type_id === 4) {
+        return i18n.t("Embedded bot");
```
This if-clause is growing large. Maybe we can use a mapping here?

abhijeetkaur replied:
Agreed!
But I don't think the number would increase a lot hereafter, as almost all bot type cases are covered.

But definitely, using a mapping would be a plus. Will work on it.

derAnfaenger commented on static/js/settings_bots.js:
```diff
@@ -185,9 +199,24 @@ exports.set_up = function () {
+
+            $('#payload_url_inputbox').hide();
+            $('#create_payload_url').removeClass('required');
         } else {
```
Maybe we can add here `else if (bot_type === GENERIC_BOT_TYPE)`? Doesn't change the logic from what I can see, but improves readability.

abhijeetkaur replied:
Yeah! It will improve readability and be easier to debug in future.

abhijeetkaur replied:
Actually it is `else if (bot_type === GENERIC_BOT_TYPE || bot_type === INCOMING_WEBHOOK)`. Since it handles the remaining 2 cases, I have simply added a comment in the code.
Makes sense?

derAnfaenger commented on zerver/views/users.py:
```diff
@@ -262,6 +263,11 @@ def add_bot_backend(request, user_profile, full_name_raw=REQ("full_name"), short
     form = CreateUserForm({'full_name': full_name, 'email': email})
+    if bot_type == UserProfile.EMBEDDED_BOT:
+        is_service_valid = any(service_name == embedded_bot_service.name for embedded_bot_service in EMBEDDED_BOTS)
+        if is_service_valid is False:
```
`if not is_service_valid`

derAnfaenger commented on zerver/worker/queue_processors.py:
```diff
@@ -485,17 +485,28 @@ def consume(self, event):
         message = cast(Dict[str, Any], event['message'])
 
+        if (event['trigger'] == 'mention'):
+            message['is_mentioned'] = True
```
If not, we should set it to false, right?

derAnfaenger commented on zerver/worker/queue_processors.py:
```diff
@@ -485,17 +485,28 @@ def consume(self, event):
+            message_content_if_bot_is_called = get_message_content_if_bot_is_called(message=message,
+                                                                                    at_mention_bot_name=user_profile.full_name)
+
+            if message_content_if_bot_is_called is not None:
```
`if message_content_if_bot_is_called` suffices :)

derAnfaenger replied:
Actually, could we just set `message['content'] = get_message_content_...` and then check if `message['content'] is not None?

abhijeetkaur replied:
There can be 2 scenarios for `message['content']` to be None:
* The bot is not called.
* The bot is called with an empty message

@neiljp is working on handling empty messages, so I have left that part untouched. Whichever PR gets merged first, the other one can make modifications accordingly. :)

abhijeetkaur commented:
@derAnfaenger Yeah! The first 3 commits are a part of my older PR #5811. That PR was opened a month back.

abhijeetkaur commented:
@derAnfaenger Thanks for the review, I have updated the last commit in accordance to your suggestions.

I have updated the first 3 commit in #5811. Since that needs to be merged before this, I will rebase this branch as and when required. :)

zulipbot-test commented:
Heads up @abhijeetkaur, we just merged some commits that conflict with the changes your made in this pull request! You can review this repository's [recent commits](https://github.com/zulip/zulip/commits/master) to see where the conflicts occur. Please rebase your feature branch against the `upstream/master` branch and [resolve](http://zulip.readthedocs.io/en/latest/git-guide.html#recover-from-a-git-rebase-failure) your pull request's merge conflicts accordingly.

timabbott commented:
Closing in favor of its rebase #7159.
